src/qutip_qip/decompositions/single_decompositions.py

```diff
@@ -0,0 +1,191 @@
+        - :math:`\textrm{B} = \textrm{R}_y \left(\frac{-\theta}{2} \right) \textrm{R}_z \left(\frac{- \left(\alpha + \beta \right)}{2} \right)`
+        - :math:`\textrm{C} = \textrm{R}_z \left(\frac{\left(-\alpha + \beta \right)}{2} \right)`
+
+
+    Parameters
+    ----------
+    input_gate : :class:`qutip.Qobj`
+        The matrix that's supposed to be decomposed should be a Qobj.
+
+    Returns
+    -------
+    tuple
+        The gates in the decomposition are returned as a tuple of :class:`Gate`
+        objects. This tuple will contain 6 elements per each :math:`1 \times 1`
+        qubit gate - 2 gates forming :math:`\textrm{A}`, 2 gates forming :math:`\textrm{B}`,
+        1 gates forming :math:`\textrm{C}`, and some global phase gate.
+    """
+    try:
+        assert num_of_qubits == 1
+    except AssertionError:
+        raise MethodError("This method is only valid for 1 qubit gates.")
+    if check_input_shape(input_gate,num_of_qubits)==True:
+        global_phase_angle = extract_global_phase(input_gate,1)
+        global_phase_angle_string = global_phase_angle/np.pi
+        input_array = cmath.rect(1,global_phase_angle)*convert_qobj_gate_to_array(input_gate)
+
+        # separate all the elements
+        a = input_array[0][0]
+        b = input_array[0][1]
+        a_star = input_array[1][1]
+        b_star = input_array[1][0] # upto this line the code could be moved to the
+        # main function and change the input parameters for _ZYZ_rotation
+        # Will be done once the issue with global phase function is corrected.
+
+        # find alpha, beta and theta
+        alpha = cmath.phase(-a_star) + cmath.phase(b_star)
+        alpha_string = alpha/np.pi # for string in circuit diagram
+        beta = cmath.phase(-a_star) - cmath.phase(b_star)
+        beta_string = beta/np.pi
+        theta = 2*np.arctan2(np.absolute(b_star), np.absolute(a))
+        theta_string = theta/np.pi
+
+
+        Phase_gate = Gate("GLOBALPHASE",targets=[0], arg_value=global_phase_angle, arg_label=r'{:0.2f} \times \pi'.format(global_phase_angle_string))
+        Rz_A = Gate("RZ",targets=[0], arg_value=alpha, arg_label=r'{:0.2f} \times \pi'.format(alpha_string))
+        Ry_A = Gate("RY",targets=[0], arg_value=theta/2, arg_label=r'{:0.2f} \times \pi'.format(theta_string/2))
+        Pauli_X = Gate("X",targets=[0])
+        Ry_B = Gate("RY",targets=[0], arg_value=-theta/2, arg_label=r'{:0.2f} \times \pi'.format(-theta_string/2))
+        Rz_B = Gate("RZ",targets=[0], arg_value=-(alpha+beta)/2, arg_label=r'{:0.2f} \times \pi'.format(-(alpha_string+beta_string)/2))
+        Rz_C = Gate("RZ",targets=[0], arg_value=(-alpha+beta)/2, arg_label=r'{:0.2f} \times \pi'.format((-alpha_string+beta_string)/2))
+
+        return(Rz_A, Ry_A, Pauli_X, Ry_B, Rz_B, Pauli_X, Rz_C, Phase_gate)
+
+    else:
+        raise GateError("Shape of input does not match with the number of qubits.")
```

Comment:
It would be nice to keep things in one place. You already raise many errors in check functions, only this one is raised here. I think it would be nice to either raise all errors there or only return `False` or `True`, never raise.

Reply:
I have tried to do so in the new changes. Let me know if I did not do what I thought was being done.